src/spaces/openstreetmap_new.jl

```diff
@@ -0,0 +1,675 @@
+    end
+    # Those not in the `go_deeper` category are searched in the positive direction
+    setdiff!(outgoing, outgoing[go_deeper])
+    to_collect = map(i -> ids_on_road(pos2, i, model), outgoing)
+    filter!(!isempty, to_collect)
+    if !isempty(to_collect)
+        for (id, dist) in vcat(to_collect...)
+            if dist <= distance
+                push!(nearby, id)
+            end
+        end
+    end
+end
+
+function search_inward_ids!(
+    nearby::Vector{Int},
+    distance::Real,
+    pos1::Int,
+    pos2::Int,
+    model,
+)
+    # find all intersections the current start position connects to
+    incoming = filter(i -> i != pos2, nearby_positions(pos1, model; neighbor_type = :in))
+    # Distances for each road
+    indist = [road_length(i, pos1, model) for i in incoming]
+    # Identify roads that are shorter than the search distance
+    go_deeper = findall(i -> i < distance, indist)
+    # Those in the `go_deeper` category are completely covered by the distance search.
+    # Any agent found in the heading or opposite directions along this road are counted.
+    for i in go_deeper
+        start = incoming[i]
+        for (id, _) in ids_on_road(start, pos1, model)
+            push!(nearby, id)
+        end
+        # We must recursively look up branches until the search distance is met
+        search_inward_ids!(nearby, distance - indist[i], start, pos1, model)
+    end
+    # Those not in the `go_deeper` category are searched in the positive direction
+    setdiff!(incoming, incoming[go_deeper])
+    to_collect = map(i -> ids_on_road(i, pos1, model, true), incoming)
+    filter!(!isempty, to_collect)
+    if !isempty(to_collect)
+        for (id, dist) in vcat(to_collect...)
+            if dist <= distance
+                push!(nearby, id)
+            end
+        end
+    end
+end
+
+function Agents.nearby_ids(
+    agent::A,
+    model::ABM{<:OpenStreetMapSpace,A},
+    args...;
+    kwargs...
+) where {A<:AbstractAgent}
+    all = nearby_ids(agent.pos, model, args...; kwargs...)
+    filter!(i -> i ≠ agent.id, all)
+end
+
+Agents.nearby_positions(pos::Tuple{Int,Int,Float64}, model, args...; kwargs...) =
+    nearby_positions(pos[1], model, args...; kwargs...)
+
+function Agents.nearby_positions(
+    position::Int,
+    model::ABM{<:OpenStreetMapSpace};
+    neighbor_type::Symbol = :default
+)
+    @assert neighbor_type ∈ (:default, :all, :in, :out)
+    neighborfn = if neighbor_type == :default
+        Graphs.neighbors
+    elseif neighbor_type == :in
+        Graphs.inneighbors
+    elseif neighbor_type == :out
+        Graphs.outneighbors
+    else
+        Graphs.all_neighbors
+    end
+    neighborfn(model.space.m.g, position)
+end
+
+end # module OSM
+
+"""
+    OpenStreetMapSpace(path::AbstractString; kwargs...)
+Create a space residing on the Open Street Map (OSM) file provided via `path`.
+The functionality related to Open Street Map spaces is in the submodule `OSM`.
+
```

Comment:
Let's please move this to the top of the file! Makes the file much easier to read if you first get the summary of the API and then see the source code!

Reply:
Sure. I didn't notice it at the bottom of the file, which I guess makes a perfect case for why it shouldn't be there